From bef19e6d5d12f9760ffd3a474b2fe5da49bfcc9c Mon Sep 17 00:00:00 2001
From: Tommy Huang <tommy_huang@aspeedtech.com>
Date: Tue, 16 Sep 2025 10:41:07 +0800
Subject: [PATCH 06/10] drivers: ast2700-mailbox: add wait bit clear in the
 mbox send

Add IPCR_STATUS waitting for claer in the mailbox sned message.

Signed-off-by: Tommy Huang <tommy_huang@aspeedtech.com>
Change-Id: If65e43a9fe57c214041f2923cf71779767681713
(cherry picked from commit a1ff4627df7399b65157d882fc18b6f9c1a2e826)
Signed-off-by: Vasanthakumar K <vasanthakumark@ami.com>
---
 drivers/mailbox/ast2700-mailbox.c | 12 ++++++++----
 1 file changed, 8 insertions(+), 4 deletions(-)

diff --git a/drivers/mailbox/ast2700-mailbox.c b/drivers/mailbox/ast2700-mailbox.c
index 042f80bb79..f3cf525594 100644
--- a/drivers/mailbox/ast2700-mailbox.c
+++ b/drivers/mailbox/ast2700-mailbox.c
@@ -43,7 +43,7 @@ static int ast2700_mbox_send(struct mbox_chan *chan, const void *data)
 	int idx = chan->id;
 	void *data_reg;
 	uint32_t *word_data;
-	int num_words;
+	int num_words, ret;
 
 	if (!(readl(mb->tx_regs + IPCR_ENABLE) & BIT(idx))) {
 		dev_warn(chan->dev, "%s: Ch-%d not enabled yet\n", __func__, idx);
@@ -66,10 +66,13 @@ static int ast2700_mbox_send(struct mbox_chan *chan, const void *data)
 	debug("%s: Ch-%d sent\n", __func__, idx);
 
 	/* Wait until observation bit is cleared */
-	wait_for_bit_le32((const void *)(mb->tx_regs + IPCR_TX_TRIG), BIT(idx), false,
-			  mb->tx_tout, false);
+	ret = wait_for_bit_le32((const void *)(mb->tx_regs + IPCR_STATUS), BIT(idx), false,
+				mb->tx_tout, false);
 
-	return 0;
+	if (ret < 0)
+		return ret;
+	else
+		return 0;
 }
 
 static int ast2700_mbox_recv(struct mbox_chan *chan, void *data)
@@ -136,6 +139,7 @@ static int ast2700_mbox_probe(struct udevice *dev)
 	data = (struct ast2700_mbox_data *)dev_get_driver_data(dev);
 	if (!data)
 		return -EINVAL;
+
 	priv->msg_size = data->msg_size;
 
 	priv->tx_tout = dev_read_u32_default(dev, "aspeed,tx-timeout", -1);
-- 
2.34.1

